staging/src/k8s.io/apimachinery/pkg/util/wait/loop.go

```diff
@@ -50,13 +50,13 @@ func loopConditionUntilContext(ctx context.Context, t Timer, immediate, sliding
 		}(); err != nil || ok {
 			return err
 		}
-	} else {
-		timeCh = t.C()
-		select {
-		case <-doneCh:
-			return ctx.Err()
-		case <-timeCh:
-		}
+	}
+
+	timeCh = t.C()
```

Comment:
@smarterclayton the problem I see know is how to handle the sliding in this first iteration

Reply:
but this change looks correct to me, just that `if !sliding` this block should be executed always before the `immediate` function and `if sliding` always after the `immediate` function

Reply:
oh, nevermind, immediate guarantees the execution, so this changes seems correct as it is now

Reply:
Hrm, I don't think this preserves sliding properly for the first iteration.

If sliding is false, then the sliding interval should start (invocation of `t.C()`) before the condition is executed (even in immediate).  If sliding is true, the sliding interval should start after the condition is executed.  Where timeCh is initialized matters (that's why it's lazily initialized in the loop).

The fact that a test case didn't fail means that we aren't properly testing for that condition.  Please add a test case that ensures sliding isn't broken in this scenario (which should then make it clear what the fix should be)?

Reply:
/lgtm cancel
@AxeZhan please take a look at Clayton´s comment